synapse/federation/sender/__init__.py

```diff
@@ -393,34 +420,59 @@ async def handle_room_events(events: Iterable[EventBase]) -> None:
+            Tuple[str, List[Tuple[EventBase, Collection[str]]]]
+        ],
+    ) -> None:
+        # Tuples of room_id + destination to their max-seen stream_ordering
+        room_with_dest_stream_ordering = {}  # type: Dict[Tuple[str, str], int]
 
-        destinations = set(destinations)
-        destinations.discard(self.server_name)
-        logger.debug("Sending to: %s", str(destinations))
+        # List of events to send to each destination
+        events_by_dest = {}  # type: Dict[str, List[EventBase]]
 
-        if not destinations:
-            return
+        # For each roomID + events with destinations pair...
+        for room_id, event_and_dests in per_room_events_and_dests:
+
+            # (dest -> stream_id)
+            this_stream_ordering = {}  # type: Dict[str, int]
+
+            # ...get every event with their destinations...
+            for event, destinations in event_and_dests:
 
-        sent_pdus_destination_dist_total.inc(len(destinations))
-        sent_pdus_destination_dist_count.inc()
+                # (we got this from the database, it's filled)
+                assert event.internal_metadata.stream_ordering
 
-        assert pdu.internal_metadata.stream_ordering
+                sent_pdus_destination_dist_total.inc(len(destinations))
+                sent_pdus_destination_dist_count.inc()
 
-        # track the fact that we have a PDU for these destinations,
-        # to allow us to perform catch-up later on if the remote is unreachable
-        # for a while.
-        await self.store.store_destination_rooms_entries(
-            destinations,
-            pdu.room_id,
-            pdu.internal_metadata.stream_ordering,
+                # ...iterate over those destinations..
+                for destination in destinations:
+                    if (
+                        this_stream_ordering.setdefault(destination, 0)
+                        < event.internal_metadata.stream_ordering
+                    ):
+                        # ...update their stream-ordering...
+                        this_stream_ordering[
+                            destination
+                        ] = event.internal_metadata.stream_ordering
+
+                    # ...and add the event to each destination queue.
+                    events_by_dest.setdefault(destination, []).append(event)
+
+            for dest, stream_id in this_stream_ordering.items():
+                room_with_dest_stream_ordering[(room_id, dest)] = stream_id
+
+        # Bulk-store destination_rooms stream_ids
+        await self.store.bulk_store_destination_rooms_entries(
+            room_with_dest_stream_ordering
         )
 
-        for destination in destinations:
-            self._get_per_destination_queue(destination).send_pdu(pdu)
+        for destination, pdus in events_by_dest.items():
+            logger.debug("Sending %d pdus to %s", len(pdus), destination)
+
+            self._get_per_destination_queue(destination).send_pdu(*pdus)
```

Comment:
my instinct is that it would be more efficient - and possibly clearer - to define `send_pdu` to take an `Iterable[EventBase]` and pass `pdus` in directly rather than using the `*` syntax. That may or may not be accurate.

Reply:
It (effectively) doesn't change the function signature, as it's still able to receive single positional items, I'd wanna change the function signature to `send_pdus` and look for other places where it is called, but every old call to it would need it's singular argument wrapped with `(arg,)` or `[arg]`, `send_pdus([arg])`, which i personally find less readable overall than using this positional argument packing.

Reply:
> but every old call to it would need it's singular argument wrapped with (arg,)

there aren't any other calls to it, except in tests...

Reply:
Alright, then i'll change the function signature :+1: